pkg/dynamicconfig/table_test.go

```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamicconfig
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/health/types"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/testutils"
+)
+
+var (
+	cmName         = "cilium-config"
+	namespace      = "kube-system"
+	dummyConfigMap = map[string]string{"key1": "value1", "key2": "value2"}
+)
+
+func TestWatchKey(t *testing.T) {
+	_, db, dct, _ := fixture(t)
+
+	done := make(chan bool)
+	key := "key"
+	lowPrioritySource := "lowPriority"
+	value := "value"
+	newValue := "newValue"
+
+	upsertDummyEntry(db, dct, key, lowPrioritySource, value)
+
+	e, w := WatchKey(db.ReadTxn(), dct, key)
+
+	if e.Key.Name != key || e.Value != value || e.Key.Source != lowPrioritySource {
+		t.Errorf("Entry mismatch for key %v: expected (key=%v, source=%v, value=%v), but got (key=%v, source=%v, value=%v)", key, key, lowPrioritySource, value, e.Key, e.Key.Source, e.Value)
+	}
+
+	go waitFor(func() {
+		<-w
+		e, _ := WatchKey(db.ReadTxn(), dct, key)
+		if e.Key.Name != key || e.Value != newValue || e.Key.Source != cmName {
+			t.Errorf("Entry mismatch for key %v: expected (key=%v, source=%v, value=%v), but got (key=%v, source=%v, value=%v)", key, key, lowPrioritySource, value, e.Key, e.Key.Source, e.Value)
+		}
+	}, done)
+
+	upsertDummyEntry(db, dct, key, cmName, newValue)
+
+	select {
+	case <-done:
+	case <-time.After(2 * time.Second):
```

Comment:
nit: I'd recommend using the assert.Eventually family of functions for this, it leads to easier to read code and tends to be more robust (although in this case this seems fine).

Reply:
Indeed, I agree, but I wanted to include the test for the watcher (chan) to make sure the index gets invalidated by the different sources.

Reply:
@ovidiutirla I don't quite understand your response here - We generally avoid timer-based waits in Cilium tests, because they're racy and can make the tests unreliable in environments with high CPU usage, which is not uncommon in CI. Is there a way to implement this test without referencing discrete time units like `time.Second`? `assert.Eventually` is a common pattern, or otherwise using channels.

Reply:
The time is used as a timeout, the channel is closed in the happy path when the values is updated in the table.

Reply:
OK. I guess if it turns out this is unreliable then we can always revisit it.